opencog/reasoning/RuleEngine/rules/pln/evaluation-to-member.scm

```diff
@@ -0,0 +1,58 @@
+; ============================================================================= 
+; GeneralEvaluationToMemberRule 
+;	Takes EvaluationLinks with >=2 Arguments and creates a Member Link
+;
+;	EvaluationLink (pred D (ListLink B C))
+;		becomes
+;	MemberLink( B (SatisfyingSetLink( Variable $X 
+;		(EvaluationLink (pred D (ListLink X C))))))
+; -----------------------------------------------------------------------------
+
+(define pln-rule-evaluation-to-member
+	(BindLink
+		(ListLink
+			(VariableNode "$A")
+			(TypedVariableLink
+				(VariableNode "$D")
+				(VariableTypeNode "PredicateNode")))
+        (ImplicationLink
+            (EvaluationLink
+                (VariableNode "$D")
+                (VariableNode "$A"))
+            (ExecutionOutputLink
+                (GroundedSchemaNode "scm:pln-formula-evaluation-to-member")
+                (ListLink
+                    (EvaluationLink
+                        (VariableNode "$D")
+                        (VariableNode "$A")))))))
+
+
+; -----------------------------------------------------------------------------
+; Evaluation To Member Formula
+; -----------------------------------------------------------------------------
+
+; -----------------------------------------------------------------------------
+; Side-effect: TruthValue of the new link stays the same
+; -----------------------------------------------------------------------------
+
+(define (pln-formula-evaluation-to-member DBC)
+	(if (= (cog-arity (gdr DBC)) 0)
+		(MemberLink (stv (cog-stv-strength DBC) (cog-stv-confidence DBC))
+			(gdr DBC)
+			(SatisfyingSetLink
+				(VariableNode "$X")
+				(EvaluationLink
+					(gar DBC)
+					(VariableNode "$X"))))
+		(MemberLink (stv (cog-stv-strength DBC) (cog-stv-confidence DBC))
+			(gadr DBC)
+			(SatisfyingSetLink
+				(VariableNode "$X")
+				(EvaluationLink
+					(gar DBC)
+					(ListLink
+						(cons
+							(VariableNode "$X")
+							(cdr (cog-get-all-nodes (gdr DBC))))))))))
```

Comment:
I'm starting to agree with you....   Implicit binding is likely to confuse
people, and right now we should value minimizing confusion over a somewhat
dubiously significant savings in memory....

LambdaLink seems a good step...


On Fri, Mar 20, 2015 at 6:42 PM, ngeiswei <notifications@github.com> wrote:

> In opencog/reasoning/RuleEngine/rules/pln/evaluation-to-member.scm
> <https://github.com/opencog/opencog/pull/1422#discussion_r26831162>:
>
> > +			(gdr DBC)
> > +			(SatisfyingSetLink
> > +				(VariableNode "$X")
> > +				(EvaluationLink
> > +					(gar DBC)
> > +					(VariableNode "$X"))))
> > +		(MemberLink (stv (cog-stv-strength DBC) (cog-stv-confidence DBC))
> > +			(gadr DBC)
> > +			(SatisfyingSetLink
> > +				(VariableNode "$X")
> > +				(EvaluationLink
> > +					(gar DBC)
> > +					(ListLink
> > +						(cons
> > +							(VariableNode "$X")
> > +							(cdr (cog-get-all-nodes (gdr DBC))))))))))
>
> Regarding the implicit vs explicit. The debate initially was whether we
> want hyperpgraphs like
>
> EvaluationLink
>     P
>     X
>
> to represent a predicate or a body function. The tradition in computer
> science has been to have a function constructor like lambda in order to
> turn a body function into a function. The function constructor allows to
> know what arguments the function takes, in which order, etc. Doing that
> implicitly I believe if risky, and then confusing for humans as well, that
> is why I suggest to use a LambdaLink instead (formerly called BindLink, but
> let's be honest, LamdbaLink is a much better name! right?). And BindLink
> can be used as the base type for all that pertains to binding, forall,
> lambda, etc.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/opencog/opencog/pull/1422/files#r26831162>.
>



-- 
Ben Goertzel, PhD
http://goertzel.org

"The reasonable man adapts himself to the world: the unreasonable one
persists in trying to adapt the world to himself. Therefore all progress
depends on the unreasonable man." -- George Bernard Shaw